microcosm_eventsource/models.py

```diff
@@ -0,0 +1,189 @@
+        ))
+
+        return super(EventMeta, cls).__new__(cls, name, bases, dct)
+
+    def make_declarations(cls, container_name, event_type, table_name):
+        """
+        Declare columns and indexes.
+
+        An event assumes the following:
+
+         -  Each event belongs to (and has a foreign key to) another "container" table.
+
+            Typically, the container table has immutable rows that define the context of an entity
+            and events track state changes to rows of this table.
+
+         -  Each event has a well-defined type, derived from an enumeration.
+
+         -  Each event has a well-defined state, consisting of one or more enumerated values.
+
+         -  Each event has an integer version, starting from one.
+
+            Not every event uses versions; those that do will generally have a uniqueness contraint
+            on some event types per version of a container row.
+
+         -  Each event has a nullable parent event id where a null value represents the first event
+            in a version and subsequent event have a unique parent id to ensure semantic ordering.
+
+         -  Each event has a non-nullable serial clock to ensure total ordering.
+
+        """
+        container_id = "{}.id".format(container_name)
+        container_id_name = "{}_id".format(container_name)
+        parent_id = "{}.id".format(table_name)
+
+        return {
+            # columns
+            container_id_name: Column(UUIDType, ForeignKey(container_id), nullable=False),
+            "event_type": Column(EnumType(event_type), nullable=False),
+            "clock": Column(Serial, server_default=FetchedValue(), nullable=False, unique=True),
+            "parent_id": Column(UUIDType, ForeignKey(parent_id), nullable=True, unique=True),
+            "state": Column(ARRAY(EnumType(event_type)), nullable=False, default=default_state),
+            "version": Column(Integer, default=1, nullable=False),
+
+            # shortcuts
+            "container_id": ColumnAlias(container_id_name),
+            "container_id_name": container_id_name,
+
+            # indexes and constraints
+            "__table_args__": cls.make_table_args(cls, container_id_name, event_type),
+        }
+
+    def make_table_args(cls, container_id_name, event_type):
+        return cls.make_indexes(
+            cls,
+            container_id_name,
+        ) + cls.make_state_machine_constraints(
+            cls,
+            event_type,
+        ) + cls.make_column_constraints(
+            cls,
+            event_type,
+        )
+
+    def make_indexes(cls, container_id_name):
+        return (
+            # logical clock is unique and indexed
+            Index(
+                "unique_logical_clock",
+                container_id_name,
+                "clock",
+                unique=True,
+            ),
+            # events are unique per container, event type, and version
+            # XXX this won't work for some types of events
```

Comment:
For my fixture example, declaring uniqueness by event type doesn't work for reassign/reschedule states. To do.